api/server/router/system/system_routes.go

```diff
@@ -32,22 +37,78 @@ func (s *systemRouter) pingHandler(ctx context.Context, w http.ResponseWriter, r
 	w.Header().Add("Cache-Control", "no-cache, no-store, must-revalidate")
 	w.Header().Add("Pragma", "no-cache")
 
-	builderVersion := build.BuilderVersion(s.features())
+	features := s.features()
+	builderVersion := build.BuilderVersion(features)
 	if bv := builderVersion; bv != "" {
 		w.Header().Set("Builder-Version", string(bv))
 	}
 
 	w.Header().Set("Swarm", s.swarmStatus())
 
+	engineFeaturesHeader, err := BuildEngineFeaturesHeader(features)
+	if err != nil {
+		return err
+	}
+	w.Header().Set("Engine-Features", engineFeaturesHeader)
+
 	if r.Method == http.MethodHead {
 		w.Header().Set("Content-Type", "text/plain; charset=utf-8")
 		w.Header().Set("Content-Length", "0")
 		return nil
 	}
-	_, err := w.Write([]byte{'O', 'K'})
+	_, err = w.Write([]byte{'O', 'K'})
 	return err
 }
 
+// The HTTP spec does not define size limits for headers, but clients usually set limits.
+// The Go default HTTP client has a default of 1MB:
+// https://cs.opensource.google/go/go/+/refs/tags/go1.23.3:src/net/http/server.go;l=916
+// but other clients and proxies usually have lower limits. To be safe, we should try to
+// set a conservative limit.
+// 100*(30 +1 (=) +1 (,) +5 (false/true)) = 3700 which is under 4K.
+const (
+	maxFeatures      = config.MaxFeatures
+	maxFeatureKeyLen = config.MaxFeatureKeyLen
+)
+
+func BuildEngineFeaturesHeader(features map[string]bool) (string, error) {
+	featuresLen := len(features)
+	if featuresLen > maxFeatures {
+		return "", errors.Errorf("too many features – expected max %d, found %d", maxFeatures, featuresLen)
+	}
+
+	// TODO(laurazard): should we still do these validations here
+	// when we're validating at the gate?
+	keys := make([]string, 0, featuresLen)
```

Comment:
Tempted to say the answer is "no". I think we should indeed validate at the gate, and for here it's probably fine to just assume they're OK 🤔

Reply:
I added a commit on top (that we can squash later if we want) moving all the validation out of the handler and into the `daemon/config` – https://github.com/moby/moby/commit/6422a88599f3ea47b6fecaca8e4a5d63f5efdaae.